tests/scatter_plot_data_utils.test.ts

```diff
@@ -0,0 +1,120 @@
+import { describe, expect, it } from "vitest";
+import { ColorRamp } from "../src/colorizer";
+import { Color } from "three";
+import {
+  TraceData,
+  getBucketIndex,
+  splitTraceData,
+  subsampleColorRamp,
+} from "../src/components/tabs/scatter_plot_data_utils";
+
+describe("subsampleColorRamp", () => {
+  it("Returns endpoints of a color ramp", () => {
+    const ramp = new ColorRamp(["#ff0000", "#0000ff"]);
+    const colors = subsampleColorRamp(ramp, 2);
+    expect(colors).to.deep.equal([new Color("#ff0000"), new Color("#0000ff")]);
+  });
+
+  it("Returns evenly spaced colors", () => {
+    const ramp = new ColorRamp(["#a0a0a0", "#000000"]);
+    const expectedColors = [
+      new Color("#a0a0a0"),
+      new Color("#808080"),
+      new Color("#606060"),
+      new Color("#404040"),
+      new Color("#202020"),
+      new Color("#000000"),
+    ];
+    const colors = subsampleColorRamp(ramp, 6);
+    expect(colors).to.deep.equal(expectedColors);
+  });
+});
+
+describe("getBucketIndex", () => {
+  const bucketSorterFactory = (minValue: number, maxValue: number, numBuckets: number) => {
+    return (value: number) => getBucketIndex(value, minValue, maxValue, numBuckets);
+  };
+
+  it("Sorts values into buckets correctly", () => {
+    // Each integer value maps to its same bucket
+    const sorter = bucketSorterFactory(0, 9, 10);
+    for (let i = 0; i < 10; i++) {
+      expect(sorter(i)).to.equal(i);
+      // Check that values between integers are also sorted correctly
+      expect(sorter(i - 0.1)).to.equal(i);
+      expect(sorter(i + 0.1)).to.equal(i);
+    }
+  });
+
+  it("Clamps values outside of range", () => {
+    const sorter = bucketSorterFactory(0, 9, 10);
+    expect(sorter(-1)).to.equal(0);
+    expect(sorter(-100)).to.equal(0);
+    expect(sorter(Number.NEGATIVE_INFINITY)).to.equal(0);
+    expect(sorter(10)).to.equal(9);
+    expect(sorter(100)).to.equal(9);
+    expect(sorter(Number.POSITIVE_INFINITY)).to.equal(9);
+  });
+
+  it("Uses rounding to determine bucket index", () => {
+    const sorter = bucketSorterFactory(0, 1, 2);
+    expect(sorter(0.499999)).to.equal(0);
+    expect(sorter(0.5)).to.equal(1);
+  });
```

Comment:
This test doesn't capture the above comment since it's sampling right in the middle of the range, but e.g.
```typescript
const sorter = bucketSorterFactory(0, 1, 4);
expect(sorter(0.2499999)).to.equal(0);
expect(sorter(0.25)).to.equal(1);
```
will fail.

Reply:
Ended up making the behavior explicit through a method comment 👍